HadesScheduler/docker/docker.go

```diff
@@ -3,103 +3,193 @@ package docker
 	"github.com/ls1intum/hades/shared/payload"
 	"github.com/ls1intum/hades/shared/utils"
-	log "github.com/sirupsen/logrus"
+	slogfluentd "github.com/samber/slog-fluentd/v2"
+	"golang.org/x/exp/maps"
 )
 
-type Scheduler struct {
-	cli                  *client.Client
-	script_executor      string
-	container_autoremove bool
-	cpu_limit            uint
-	memory_limit         string
-}
-
-type DockerConfig struct {
+type DockerEnvConfig struct {
 	DockerHost           string `env:"DOCKER_HOST" envDefault:"unix:///var/run/docker.sock"`
 	ContainerAutoremove  bool   `env:"DOCKER_CONTAINER_AUTOREMOVE" envDefault:"true"`
 	DockerScriptExecutor string `env:"DOCKER_SCRIPT_EXECUTOR" envDefault:"/bin/bash -c"`
 	CPU_limit            uint   `env:"DOCKER_CPU_LIMIT"`    // Number of CPUs - e.g. '6'
 	MEMORY_limit         string `env:"DOCKER_MEMORY_LIMIT"` // RAM usage in g or m  - e.g. '4g'
 }
 
-func NewDockerScheduler() Scheduler {
-	var dockerCfg DockerConfig
+type DockerProps struct {
+	scriptExecutor       string
+	containerAutoremove  bool
+	cpu_limit            uint
+	memory_limit         string
+	volumeName           string
+	containerLogsOptions container.LogConfig
+}
+
+type Scheduler struct {
+	cli *client.Client
+	DockerProps
+	fluentd.FluentdOptions
+}
+
+type DockerJob struct {
+	cli    *client.Client
+	logger *slog.Logger
+	DockerProps
+	payload.QueuePayload
+}
+
+type DockerStep struct {
+	cli    *client.Client
+	logger *slog.Logger
+	DockerProps
+	payload.Step
+}
+
+func NewDockerScheduler() *Scheduler {
+	var dockerCfg DockerEnvConfig
 	utils.LoadConfig(&dockerCfg)
-	log.Debugf("Docker config: %+v", dockerCfg)
+	slog.Debug("Docker config: %+v", "config", dockerCfg)
 
 	var err error
 	// Create a new Docker client
 	cli, err := client.NewClientWithOpts(client.WithHost(dockerCfg.DockerHost), client.WithAPIVersionNegotiation())
 	if err != nil {
-		log.WithError(err).Fatal("Failed to create Docker client")
+		slog.Error("Failed to create Docker client", slog.Any("error", err))
 	}
-	return Scheduler{
-		cli:                  cli,
-		container_autoremove: dockerCfg.ContainerAutoremove,
-		script_executor:      dockerCfg.DockerScriptExecutor,
-		cpu_limit:            dockerCfg.CPU_limit,
-		memory_limit:         dockerCfg.MEMORY_limit,
+	return &Scheduler{
+		cli: cli,
+		DockerProps: DockerProps{
+			scriptExecutor:      dockerCfg.DockerScriptExecutor,
+			containerAutoremove: dockerCfg.ContainerAutoremove,
+			cpu_limit:           dockerCfg.CPU_limit,
+			memory_limit:        dockerCfg.MEMORY_limit,
+		},
+	}
+}
+
+func (d *Scheduler) SetFluentdLogging(addr string, max_retries uint) *Scheduler {
+	if addr != "" {
+		d.FluentdOptions = fluentd.FluentdOptions{
+			Addr:     addr,
+			MaxRetry: max_retries,
+		}
 	}
+	return d
 }
 
 func (d Scheduler) ScheduleJob(ctx context.Context, job payload.QueuePayload) error {
+	// Create a custom logger only for this job
+	var job_logger *slog.Logger = slog.Default()
```

Comment:
Is this a good idea? I'd rather not assign the default logger here. (Especially with the comment above). What was the rational behind this?

Reply:
I instantiate the default logger here in case there is no `fluentd` address.
Only if `fluentd` is configure (`if d.FluentdOptions.Addr != ""`), I will create a dedicated one for the job, otherwise I will just use the default one.

Does that make sense to you?

Reply:
I see the point - Still i think this may lead to hard to debug problems in case the "specific" logger is not configured correctly. I think it would be better to default to the standard logger in the else branch and also inform the user that we default to the standard. Any strong opinions on that?